Fix MyPy type errors in models/errors.py & connection.py

RoyLee1224 opened the pull request:
## Related issue

part of #56735

## Changes

Added validation in `Connection.__init__`. It now allows `conn_id` to be `None` only for URI-based connections (used`type: ignore` just for that specific path).
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

RoyLee1224 committed:
fix: Fix MyPy type errors in errors.py & connection.py

RoyLee1224 committed:
fix: Fix mypy errors in Connection model init

vincbeck commented on airflow-core/src/airflow/models/connection.py:
```diff
@@ -158,7 +158,7 @@ def __init__(
     ):
         super().__init__()
-        self.conn_id = sanitize_conn_id(conn_id)
+        self.conn_id = sanitize_conn_id(conn_id)  # type: ignore[assignment]
```
We should try ti solve it instead of ignoring it

RoyLee1224 replied:
Sure, I'm on it

vincbeck commented on airflow-core/src/airflow/models/connection.py:
```diff
@@ -128,8 +128,8 @@ class Connection(Base, LoggingMixin):
-    conn_id: Mapped[str] = mapped_column(String(ID_LEN), unique=True, nullable=False)
-    conn_type: Mapped[str] = mapped_column(String(500), nullable=False)
+    conn_id: Mapped[str | None] = mapped_column(String(ID_LEN), unique=True, nullable=False)
+    conn_type: Mapped[str | None] = mapped_column(String(500), nullable=False)
```
Those are not consistent. Should they be nullable? If yes then the type should be `str | None` and `nullable=True`, if not it should be `str` and `nullable=False`

RoyLee1224 replied:
Sure, I'll add runtime validation in __init__ and fix the unit tests.

RoyLee1224 committed:
fix: Enforce non-null conn_id in Connection model and update tests

RoyLee1224 committed:
Fix 17 test files to provide required conn_id and conn_type parameters

RoyLee1224 committed:
Fix tests by providing required conn_id and conn_type parameters

RoyLee1224 committed:
fix: Revert test modifications

RoyLee1224 committed:
fix: Fix MyPy type errors in Connection model init

  - Add conditional assignment for conn_id to handle sanitize_conn_id() returning None
  - Add conditional assignment for conn_type to handle None parameter values
  - Change get_uri() return type from str to str | None for type safety
  - Add explicit type annotations for host_to_use and protocol_to_add variables
  - Fix connection URI caching to handle None return from get_uri()

RoyLee1224 committed:
fix: Add conditional assignment for conn_id to handle sanitize_conn_id() returning None

RoyLee1224 committed:
refactor: De-duplicate conn_id logic

jason810496 reviewed (approved):
Nice!

vincbeck reviewed (approved):
Nice!
